Address Safer CPP warnings in HTMLDetailsElement.cpp and HTMLSummaryElement.cpp

nt1m opened the pull request:
#### 8bbba2aa777cf326e24a4e5c63b1f51bd52315ff
<pre>
Address Safer CPP warnings in HTMLDetailsElement.cpp and HTMLSummaryElement.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=289225">https://bugs.webkit.org/show_bug.cgi?id=289225</a>
<a href="https://rdar.apple.com/146363831">rdar://146363831</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::Node::protectedShadowRoot const):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::protectedTarget const):
* Source/WebCore/dom/Event.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::DetailsSlotAssignment::slotNameForHostChild const):
(WebCore::HTMLDetailsElement::didAddUserAgentShadowRoot):
(WebCore::HTMLDetailsElement::isActiveSummary const):
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
(WebCore::HTMLDetailsElement::otherElementsInNameGroup):
* Source/WebCore/html/HTMLSummaryElement.cpp:
(WebCore::HTMLSummaryElement::detailsElement const):
(WebCore::HTMLSummaryElement::defaultEventHandler):

Canonical link: <a href="https://commits.webkit.org/291716@main">https://commits.webkit.org/291716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64128f6d5c0a30bf08e0a90e997ee74c4505d004

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93796 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/13370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3109 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13666 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21812 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/98802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96798 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2391 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/43637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2462 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100837 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20848 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/21100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80708 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/79954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19897 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15043 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20832 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/23979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/22260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,53 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
 void HTMLDetailsElement::didAddUserAgentShadowRoot(ShadowRoot& root)
 {
-    auto summarySlot = HTMLSlotElement::create(slotTag, document());
+    auto summarySlot = HTMLSlotElement::create(slotTag, protectedDocument());
```
Instead of keep calling protectedDocument() 5 times, declare a `RefPtr` local variable and store `document()` there.

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
 
-    auto defaultSummary = HTMLSummaryElement::create(summaryTag, document());
-    defaultSummary->appendChild(Text::create(document(), defaultDetailsSummaryText()));
+    auto defaultSummary = HTMLSummaryElement::create(summaryTag, document);
```
`auto` -> `Ref`

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
 
-    m_defaultSlot = HTMLSlotElement::create(slotTag, document());
-    m_defaultSlot->setUserAgentPart(UserAgentParts::detailsContent());
+    auto defaultSlot = HTMLSlotElement::create(slotTag, document);
```
`auto` -> `Ref`

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
+    defaultSlot->setInlineStyleProperty(CSSPropertyContentVisibility, CSSValueHidden);
+    defaultSlot->setInlineStyleProperty(CSSPropertyDisplay, CSSValueBlock);
+    root.appendChild(defaultSlot);
+    m_defaultSlot = defaultSlot.ptr();
```
`= WTFMove(defaultSlot);` would be more efficient.

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
 
     static MainThreadNeverDestroyed<const String> stylesheet(StringImpl::createWithoutCopying(detailsElementShadowUserAgentStyleSheet));
-    auto style = HTMLStyleElement::create(HTMLNames::styleTag, document(), false);
+    auto style = HTMLStyleElement::create(HTMLNames::styleTag, document, false);
```
`auto` -> `Ref`

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
     if (!slot)
         return false;
-    return slot == m_summarySlot.get();
+    return slot == summarySlot.get();
```
We could remove the `if` check above and do:
```
return slot && slot == summarySlot.get();
```

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -103,50 +103,54 @@ HTMLDetailsElement::~HTMLDetailsElement() = default;
         return false;
 
-    RefPtr slot = shadowRoot()->findAssignedSlot(summary);
+    RefPtr root = shadowRoot();
```
We should not add this local and call `protectedShadowRoot()` instead.

cdumez commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -190,9 +195,9 @@ Vector<RefPtr<HTMLDetailsElement>> HTMLDetailsElement::otherElementsInNameGroup(
-            otherElementsInNameGroup.append(&element);
+    for (Ref element : descendantsOfType<HTMLDetailsElement>(rootNode())) {
+        if (element.ptr() != this && element->attributeWithoutSynchronization(nameAttr) == detailElementName)
+            otherElementsInNameGroup.append(element.ptr());
```
`WTFMove(element)` would be more efficient.

cdumez commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -94,7 +94,8 @@ void HTMLSummaryElement::defaultEventHandler(Event& event)
     if (isActiveSummary()) {
         auto& eventNames = WebCore::eventNames();
-        if (event.type() == eventNames.DOMActivateEvent && !isInSummaryInteractiveContent(event.target())) {
+        RefPtr target = event.target();
```
We don't need this local variable, we should call `protectedTarget()` below instead.

nt1m committed:
Address Safer CPP warnings in HTMLDetailsElement.cpp and HTMLSummaryElement.cpp
https://bugs.webkit.org/show_bug.cgi?id=289225
rdar://146363831

Reviewed by Ryosuke Niwa.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::Node::protectedShadowRoot const):
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::protectedTarget const):
* Source/WebCore/dom/Event.h:
* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::DetailsSlotAssignment::slotNameForHostChild const):
(WebCore::HTMLDetailsElement::didAddUserAgentShadowRoot):
(WebCore::HTMLDetailsElement::isActiveSummary const):
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
(WebCore::HTMLDetailsElement::otherElementsInNameGroup):
* Source/WebCore/html/HTMLSummaryElement.cpp:
(WebCore::HTMLSummaryElement::detailsElement const):
(WebCore::HTMLSummaryElement::defaultEventHandler):

Canonical link: https://commits.webkit.org/291716@main
